type-check.ts

```diff
@@ -203,34 +218,78 @@ export function tcStmt(env : GlobalTypeEnv, locals : LocalTypeEnv, stmt : Stmt<n
  * @param value Type of the value passed into this destructure
  */
 function tcDestructure(env: GlobalTypeEnv, locals: LocalTypeEnv, destruct: Destructure<null>, value: Type): Destructure<Type> {
-  let values: Type[];
-  
-  if (destruct.isDestructured) {
-    throw new TypeCheckError("Destructured assignment not implemented");
-  } else if (destruct.targets.length === 1) {
-    values = [value];
+
+  if (!destruct.isDestructured) {
+    let target = destruct.targets[0];
+    return {
+      isDestructured: false,
+      targets: [{
+        target: tcAssignable(env, locals, target.target),
+        starred: target.starred,
+        ignore: target.ignore
+      }]
+    };
+  }
+
+  let types: Type[] = [];
+  if (value.tag === "class") { // This is a temporary hack to get destructuring working (reuse for tuples later?)
+    let cls = env.classes.get(value.name);
+    if (cls === undefined)
+      throw new Error(`Class ${value.name} not found in global environment. This is probably a parsing bug.`);
+    let attrs = cls[0];
+    attrs.forEach(val => types.push(val));
+  } else if (value.tag === "number" || value.tag === "bool" || value.tag === "none") {
+    types = [value];
   } else {
-    throw new TypeCheckError(`Non-destructured assignment must have 1 target. Got ${destruct.targets.length}`);
+    throw new TypeCheckError(`Type ${value.tag} cannot be destructured`);
   }
 
-  const tTargets = destruct.targets.map(({target, ignore, starred}, i) => {
```

Comment:
I'm curious, what was the logic for completely getting rid of this `.map` here and fully replacing with multiple index based for loops? was there something in particular this could not handle at all?

Reply:
`.map` couldn't handle starred targets, since it wouldn't know how far to "skip ahead" before assigning to the rest of the non-starred targets.

Reply:
```typescript
let starredOffset = 0;
map(() => {
  valType = values[i + starredOffset]
})
```
I get what you're saying in general though